tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,390 @@
+		ti = ts_scan_iterator_tuple_info(&iterator);
+		tuple = ts_scan_iterator_fetch_heap_tuple(&iterator, false, &should_free);
+		form = (Form_continuous_aggs_materialization_invalidation_log) GETSTRUCT(tuple);
+
+		/* Entry is completely enclosed by the refresh window */
+		if (form->lowest_modified_value >= refresh_window->start &&
+			form->greatest_modified_value < refresh_window->end)
+		{
+			/*
+			 * Entry completely enclosed so can be deleted:
+			 *
+			 * |---------------|
+			 *     [+++++]
+			 */
+			ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+			ts_catalog_delete_tid_only(ti->scanrel, ts_scanner_get_tuple_tid(ti));
+			ts_catalog_restore_user(&sec_ctx);
+		}
+		else if (form->lowest_modified_value < refresh_window->start &&
+				 form->greatest_modified_value >= refresh_window->end)
+		{
+			HeapTuple newtup;
+			/*
+			 * Need to cut in both ends:
+			 *
+			 *     |------|
+			 * [+++++++++++++]
+			 *
+			 * [++]        [+]
+			 */
+			update_tuple = create_invalidation_tup(tuple,
+												   form->lowest_modified_value,
+												   refresh_window->end - 1);
+			newtup =
+				create_invalidation_tup(tuple, form->greatest_modified_value, refresh_window->end);
+
+			ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+			ts_catalog_insert_only(ti->scanrel, newtup);
+			ts_catalog_restore_user(&sec_ctx);
+			heap_freetuple(newtup);
+		}
+		else if (form->lowest_modified_value < refresh_window->start &&
+				 form->greatest_modified_value >= refresh_window->start &&
+				 form->greatest_modified_value < refresh_window->end)
+		{
+			/*
+			 * Need to cut in right end:
+			 *
+			 *     |------|
+			 * [++++++]
+			 *
+			 * [++]
+			 */
+			update_tuple = create_invalidation_tup(tuple,
+												   form->lowest_modified_value,
+												   refresh_window->end - 1);
+		}
+		else if (form->lowest_modified_value >= refresh_window->start &&
+				 form->lowest_modified_value < refresh_window->end &&
+				 form->greatest_modified_value >= refresh_window->end)
+		{
+			/*
+			 * Need to cut in left end:
+			 *
+			 * |------|
+			 *    [++++++++]
+			 *
+			 *         [+++]
+			 */
+			update_tuple = create_invalidation_tup(tuple,
+												   refresh_window->start,
+												   form->greatest_modified_value);
+		}
+
+		if (update_tuple)
+		{
+			ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+			ts_catalog_update_tid_only(ti->scanrel, ts_scanner_get_tuple_tid(ti), update_tuple);
+			ts_catalog_restore_user(&sec_ctx);
+			heap_freetuple(update_tuple);
+		}
+		MemoryContextSwitchTo(oldmctx);
+		MemoryContextReset(state->per_tuple_mctx);
+	}
+
+	ts_scan_iterator_close(&iterator);
+	UnregisterSnapshot(iterator.ctx.snapshot);
+}
+
+static void
+continuous_agg_invalidation_exec(CaggInvalidationState *state,
+								 const InternalTimeRange *refresh_window)
+{
+	move_invalidations_from_hyper_to_cagg_log(state);
```

Comment:
No table level locks before update/deleting the data  in the invalidation log tables. I strongly suggest acquiring and releasing locks in this function so that it is explicit without sifting through all the function calls.

Reply:
I actually found the lock for one of the tables. But not the other table?